### ISSUE TITLE
CRITICAL BUG in new tracking implementation (v2.0.1)

### ISSUE THREAD

**Issue by PaulSchiretz** [woocommerce/woocommerce-google-analytics-integration#382]
### Describe the bug:

When a Product is added or removed from Cart, the tracking script relies on the add_to/remove_from cart button of that product_id to be part of the current DOM, otherwise a JS error is triggered.

COMMON WAYS THIS ASSUMTION IS WRONG:
Grouped, Bundled, Chained Products, BOGO Deals, Free Products you name it..... which Plugin is used or if it is done by code doesn't matter, whenever a product where the add_to/remove_from cart button is not part of the DOM, is added/removed from cart we get a major crash.

This is extremely critical as it breaks compatibility with 100s of extensions, even the ones from Automattic/Woocommerce itself.

There is no way to avoid the added_to_cart/remove_from_cart events to trigger other than editing your js files. As the event handlers are hardcoded in js and recieve events EVEN IF tracking added_to_cart/remove_from_cart events is disabled via the settings screen.

### Steps to reproduce:
1. Setup Woocommerce
2. Install ANY Extension from Automattic/Woocommerce or a 3rd Party vendor that adds Products to cart or removes them, that don't rely on an add_to_cart or remove_from_cart button.

If you want to test it via a Plugin, here are some Examples:
Woocommerces build-in Grouped Products
Product Bundles by WooCommerce
Composite Products by WooCommerce
Any BOGO Deals / Coupons Plugin by Woocommerce

4. Trigger add_to_cart/remove_from_cart on that product not part of the current html markup.

ALTERNATIVE

1. Setup Woocommerce
2. trigger add_to_cart or remove_from_cart on a product that is not part of the current html markup.

### Expected behavior:
1. The add_to_cart/remove from cart event is triggered as usual with the product_id. Or if the product id is not found on that page the event is simply ignored.

2. if added_to_cart/remove_from_cart events is disabled via the settings screen the classic.js implementation should not listen to them.

### Actual behavior:
JS error and breakage of the page.
The crash happens in any of the functions of assets/js/src/utils/index.js, cause the passed product is null/undefined.
Why is it null, cause it is not part of the products array in assets/js/src/integrations/classic.js, that is used to fetch the product.

BUT: The actual issue is that the plugin hooks into add_to_cart/remove_from_cart actions and triggering all the js events and relying on the product add/remove button being part of the DOM, which is not the case in a number of cases as explained before.

### Additional details:
We really need to get this fixed cause right now, nearly every woocommerce site out there is broken caused by the change in tracking.

In short: what you indent to track via js is add_to/remove_from cart button clicks, but what you are actually tracking is add_to/remove_from cart events - via woocommerce hooks - which is a total different thing :-/ so you cannot rely on a button being in DOM.

I provided a quick solution in the comment below! would be nice if someone @tomalec @puntope @martynmjones can have a look at this!

**Comment by PaulSchiretz** [comment]
As a quick solution to this issue, i would propose adding checks if the product is valid to:
assets/js/src/tracker/data-formatting.js

just like you already do in view_item:

```
if ( ! product ) {
	return false;
}
```

why not add it to add_to_cart, remove_from_cart, select_content, trackChangeCartItemQuantity as well:

```


/**
 * Formats data for the add_to_cart event
 *
 * @param {Object} params              The function params
 * @param {Array}  params.product      The product to track
 * @param {number} [params.quantity=1] The quantity of that product in the cart.
 */
export const add_to_cart = ( { product, quantity = 1 } ) => {

	/*ADDED*/
	if ( ! product ) {
		return false;
	}
	/******/
	return {
		items: [ getProductFieldObject( product, quantity ) ],
	};
};

/**
 * Formats data for the remove_from_cart event
 *
 * @param {Object} params              The function params
 * @param {Array}  params.product      The product to track
 * @param {number} [params.quantity=1] The quantity of that product in the cart.
 */
export const remove_from_cart = ( { product, quantity = 1 } ) => {
	/*ADDED*/
	if ( ! product ) {
		return false;
	}
	/******/
	return {
		items: [ getProductFieldObject( product, quantity ) ],
	};
};

/**
 * Tracks change_cart_quantity event
 *
 * @param {Object} params              The function params
 * @param {Array}  params.product      The product to track
 * @param {number} [params.quantity=1] The quantity of that product in the cart.
 */
export const trackChangeCartItemQuantity = ( { product, quantity = 1 } ) => {
	/*ADDED*/
	if ( ! product ) {
		return false;
	}
	/******/
	trackEvent( 'change_cart_quantity', {
		event_category: 'ecommerce',
		event_label: __(
			'Change Cart Item Quantity',
			'woocommerce-google-analytics-integration'
		),
		items: [ getProductFieldObject( product, quantity ) ],
	} );
};

/**
 * Formats data for the select_content event.
 *
 * @param {Object} params         The function params
 * @param {Object} params.product The product to track
 */
export const select_content = ( { product } ) => {
	/*ADDED*/
	if ( ! product ) {
		return false;
	}
	/******/
	return {
		content_type: 'product',
		content_id: getProductId( product ),
	};
};
```

and everything is usable again!

I would really appreciate this fix to be pushed, sorry for not creating a pull request and to bother you so much and to be a bit pushy ;-)

**Comment by mikkamp** [comment]
Hi @PaulSchiretz 

Thanks a lot for such a detailed report, and providing suggestions on how to resolve the issue.

We'd be happy to incorporate the additional check, although before we do so I'd like to be clear exactly how to reproduce the scenario, where those functions are called without a product (in particular the `add_to_cart` and `remove_from_cart` functions mentioned above). I did some testing with default themes like Storefront/Twenty Twenty and haven't been able to reproduce the error even when trying with grouped / bundled products.

You mentioned:
> When a Product is added or removed from Cart, the tracking script relies on the add_to/remove_from cart button of that product_id to be part of the current DOM, otherwise a JS error is triggered.

The product data for the relevant product does need to be available, but this isn't only searched for in the DOM. We also have cases where we add the product data through the script `woocommerce-google-analytics-integration-data-js-after`. You also mentioned that tracking events can't be disabled (that works here locally). The configuration for which events are enabled is determined in the script `woocommerce-google-analytics-integration-js-before`. Is there a possibility that those inline JS snippets are being cached, which is preventing them from having the accurate data for both the products and the enabled events?

Could you provide some more details of a page where you are seeing the error and check (in the page source), what data is being sent through the inline scripts `woocommerce-google-analytics-integration-data-js-after` and `woocommerce-google-analytics-integration-js-before`.

**Comment by PaulSchiretz** [comment]
Hi @mikkamp,

1. Maybe got to mention here amost forgot, me personally i add the (bundeled) product via ajax, maybe thats whats causing the issue
Cause of course then:

> We also have cases where we add the product data through the script woocommerce-google-analytics-integration-data-js-after.

will not be added to the code, as well as possibly some other things, maybe thats also an issue.

2. Nope, no cache no nothing just a plain empty dev machine with woocommerce installed.

3. Eather way i think the checks wouldn't hurt and solve the issue in every possible scenario and prevent crashes all over the place, i guess for a lot of folks.

Let me now if you need more info!

One more thing:

I guess that is also something somehow connected:
[https://github.com/woocommerce/woocommerce-google-analytics-integration/issues/377#issuecomment-1986852959](https://github.com/woocommerce/woocommerce-google-analytics-integration/issues/377#issuecomment-1986852959)


Eather way, i would really appreciate if we could add the checks and i can have a try with an updated version and let you know.
:-D I'm currently at work doing c++ coding, but i would be able to test a "beta" or real release in the afternoon, if you could provide one.

All the best,
Paul

**Comment by PaulSchiretz** [comment]
just to add some more reports of this bug here where you can also see the js error that would be prevented:

[https://wordpress.org/support/topic/js-error-on-product-page-2/](https://wordpress.org/support/topic/js-error-on-product-page-2/)

Exactly what i was seeing before i started to investigate...

**Comment by PaulSchiretz** [comment]
Hi @mikkamp 

Ok , i ran some more tests and, the issue happens only via AJAX!!! Cause then the product information is missing in the page!

Just tested this quickly via WPC Product Bundles for WooCommerce + WPC AJAX Add to Cart for WooCommerce thats just my test setup for now. Get the error all the time... on single product pages! i guess you need some ajax plugin(also tested it with my own implementation) + A plugin that adds a product currently not visible on a page.

This worked fine with all version prior to 2.0, and will be fixed by the additional checks if the product is valid.

Would still be nice to have the fix in the plugin! :-)

**Comment by mikkamp** [comment]
> the issue happens only via AJAX

Ah thanks for clarifying that does help a lot. I had already tried to reproduce a scenario similar to yours by adding by URL:
`https://domain.test/product/first-product-name/?add-to-cart=28`

That basically does the same as the code snippet you suggested, but I thought it might handle it differently so I setup all kinds of tests. Since you kept referring back to the function `document.body.onadded_to_cart` I was assuming it had to do something with AJAX. I was able to reproduce the same with a similar plugin then the one you suggested, I used the plugin https://wordpress.org/plugins/woo-ajax-add-to-cart/

In the forum thread you linked to the error was triggered from the script `ajax-add-to-cart-button-public.js`.
That's not a standard WC script so I was assuming there was something else involved.

Either way we are going to take a look to see what we can do to support that scenario, just returning early is a quick fix, but in essence it's just turning off tracking there.

**Comment by PaulSchiretz** [comment]
Ok, i really appreciate this! :-)

Yes i know it's a quick fix, but for now it would be better than, turning off tracking overall or have all kind of js errors.
So i would still welcome it, so i can add google analytics for woocommerce again in my production setup! :-) and used all the great consent stuff and hooks you provide with the new version! :-)

Eitherway, looking forward to a solution or a quick fix for this!

Thanks @mikkamp for your patients and great work!
